### PR TITLE
feat(gateway): multi-zone cross-mesh MeshGateway

### PR DESCRIPTION
**Pull request by michaelbeaumont**
### Summary

- Uses `kuma.io/mesh` to mark an available service as being available from other meshes

~Blocked on https://github.com/kumahq/kuma/pull/4439~

### CODE REVIEW

File: pkg/xds/ingress/dataplane.go [hunk 1]
```diff
@@ -91,6 +111,66 @@ func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService,
...
+	endpointMap := topology.BuildCrossMeshEndpointMap(
+		mesh, nil, "", meshGateways, dataplanes, nil, nil,
+	)
+	tagSets := tagSets{}
+	for _, endpoints := range endpointMap {
+		for _, endpoint := range endpoints {
+			meshName := mesh.GetMeta().GetName()
+			tagSets.addInstanceOfTags(meshName, envoy.Tags(mesh_proto.Merge(
+				map[string]string{
+					mesh_proto.MeshTag: mesh.GetMeta().GetName(),
```
Comment (lobkovilya) [inline pkg/xds/ingress/dataplane.go:165]: there is already `meshName` variable

File: pkg/xds/ingress/dataplane.go [hunk 2]
```diff
@@ -91,6 +111,66 @@ func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService,
...
+			Dataplanes: dataplanes,
+		})
+	}
+
+	return meshGatewayDataplanes, nil
+}
+
+func getIngressAvailableMeshGateways(mesh *core_mesh.MeshResource, meshGateways []*core_mesh.MeshGatewayResource, dataplanes []*core_mesh.DataplaneResource) []*mesh_proto.ZoneIngress_AvailableService {
+	// We just use this to get the meshGateways to add them to availableservices
+	endpointMap := topology.BuildCrossMeshEndpointMap(
```
Comment (lobkovilya) [inline pkg/xds/ingress/dataplane.go:156]: I might be wrong, but it seems like if you create a separate method to get the meshGateways, you don't need a full `mesh` resource, only its name. This means you don't have to make a request to resourceManager in `getMeshGateways`

File: pkg/xds/sync/ingress_proxy_builder.go [hunk 3]
```diff
@@ -57,7 +57,13 @@ func (p *IngressProxyBuilder) Build(key core_model.ResourceKey) (*xds.Proxy, err
 	if err != nil {
 		return nil, err
 	}
-	routing := p.resolveRouting(zoneIngress, zoneEgressesList, allMeshDataplanes, availableExternalServices)
+
+	allMeshGateways := &core_mesh.MeshGatewayResourceList{}
```
Comment (lobkovilya) [inline pkg/xds/sync/ingress_proxy_builder.go:61]: Can you use these resources inside the `buildZoneIngressProxy` method, instead of fetching them again?

Comment (michaelbeaumont) [reply pkg/xds/sync/ingress_proxy_builder.go:61]: I think the result is cached but yeah I switch the order around.

File: pkg/plugins/runtime/k8s/webhooks/secret_validator.go [hunk 4]
```diff
@@ -12,16 +12,13 @@ import (
...
 
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
 	secret_manager "github.com/kumahq/kuma/pkg/core/secrets/manager"
 	"github.com/kumahq/kuma/pkg/core/validators"
 	common_k8s "github.com/kumahq/kuma/pkg/plugins/common/k8s"
 	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
 )
 
-const (
-	meshLabel = "kuma.io/mesh"
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply pkg/plugins/runtime/k8s/webhooks/secret_validator.go:22]: Good point. I'll separate them out again and keep the status quo

File: test/e2e_env/multizone/gateway/cross-mesh.go [hunk 5]
```diff
@@ -99,29 +115,55 @@ func CrossMeshGatewayOnMultizone() {
...
+				Expect(egress.ResetCounters()).To(Succeed())
+
+				k8s_gateway.SuccessfullyProxyRequestToGateway(
+					*zone, gatewayMesh,
+					gatewayAddr,
+					gatewayClientNamespaceOtherMesh,
+				)
+
+				Expect(egress.GetStats(filter)).To(stats.BeGreaterThanZero())
+			})
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/multizone/gateway/cross-mesh.go:166]: aren't those test cases exactly the same?

Comment (michaelbeaumont) [reply test/e2e_env/multizone/gateway/cross-mesh.go:166]: Fixed

**Comment by jakubdyszkiewicz** [comment]
Can I ask to avoid force pushing on the branch? It makes review harder because I cannot see what exactly changed.
I'd prefer to do merges and commits with fixes, they are squashed anyways at the end.

**Comment by michaelbeaumont** [comment]
@jakubdyszkiewicz You can see what changed by clicking on "force-pushed" but yeah this one should have been a new commit indeed, once the PR starts being reviewed it should be commits in most cases

**Review by jakubdyszkiewicz — APPROVED** [review]

**Review by lobkovilya — APPROVED** [review]

**Pull request merged**